src/system_reader.rs

```diff
@@ -542,10 +630,39 @@ impl<R: RawReader> IonReader for SystemReader<R> {
                 self.lst.has_found_imports = false;
                 new_lst_state = NotReadingAnLst;
             }
-            ProcessingLstImports | ProcessingLstSymbols | ProcessingLstOpenContent => {
+            ProcessingLstSymbols | ProcessingLstOpenContent => {
+                // We're inside one of the LST fields. Finish processing the current level before
+                // stepping out.
+                self.finish_reading_current_level()?;
+                // If the upcoming call to step_out() will cause us to leave the field altogether,
+                // update our state to indicate that we're back at depth=1.
+                if self.depth() == 2 {
+                    new_lst_state = BetweenLstFields;
+                }
+            }
+            ProcessingLstImports => {
                 // We're inside one of the LST fields. Finish processing the current level before
                 // stepping out.
                 self.finish_reading_current_level()?;
+                // Finish processing LST shared symbol table import by retrieving shared symbol table from catalog.
+                if self.lst.current_import_name.is_some()
+                    && self.lst.current_import_version.is_some()
+                {
+                    if let Some(catalog) = self.catalog.as_ref() {
```

Comment:
What if `finish_reading_current_level()` passes many different structs within `imports`? This looks like it expects a single import struct.

`finish_reading_current_level()` visits every import struct in the list -- shouldn't that be enough to process each struct?

Reply:
Yes, this has been changed with the newly added LST import read states.